src/pymmcore_plus/mda/_engine.py

```diff
@@ -44,21 +51,65 @@ def setup_event(self, event: MDAEvent) -> None:
             x = event.x_pos if event.x_pos is not None else self._mmc.getXPosition()
             y = event.y_pos if event.y_pos is not None else self._mmc.getYPosition()
             self._mmc.setXYPosition(x, y)
+
         if event.z_pos is not None:
-            self._mmc.setZPosition(event.z_pos)
+            p_idx = event.index.get("p", None)
+            if p_idx not in self._z_correction:
+                self._z_correction[p_idx] = 0.0
+            self._mmc.setZPosition(event.z_pos + self._z_correction[p_idx])
+
         if event.channel is not None:
             self._mmc.setConfig(event.channel.group, event.channel.config)
         if event.exposure is not None:
             self._mmc.setExposure(event.exposure)
 
         self._mmc.waitForSystem()
 
-    def exec_event(self, event: MDAEvent) -> Any:
-        """Execute an individual event and return the image data."""
-        # TODO: add non-aquisition event-specific logic here later
-        self._mmc.snapImage()
-        return EventPayload(image=self._mmc.getImage())
+    def exec_event(self, event: MDAEvent) -> tuple[ndarray, MDAEvent] | None:
+        """Execute an individual event."""
+        action = event.action if hasattr(event, "action") else AcquireImage()
```

Comment:
change this to `action = getattr(event, 'action', None)` ... then don't create an empty `AcquireImage()` thing... just keep the same bit at the end that will hit of none of the other action instance checks hit 
```py
self._mmc.snapImage()
return EventPayload(image=self._mmc.getImage())
```